src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/HostingDiagnosticListener.cs

```diff
@@ -360,17 +480,26 @@ private RequestTelemetry InitializeRequestTelemetry(HttpContext httpContext, Act
                 activity.UpdateTelemetry(requestTelemetry, false);
             }
 
-            foreach (var prop in activity.Baggage)
+            if (this.configuration != null && !string.IsNullOrEmpty(requestTelemetry.Context.Operation.Id)
+                && SamplingScoreGenerator.GetSamplingScore(requestTelemetry.Context.Operation.Id) >= this.configuration.GetLastObservedSamplingPercentage(requestTelemetry.ItemTypeFlag))
+            {
+                requestTelemetry.IsProactivelySampledOut = true;
+            }            
+
+            if (!requestTelemetry.IsProactivelySampledOut)
             {
-                if (!requestTelemetry.Properties.ContainsKey(prop.Key))
+                foreach (var prop in activity.Baggage)
                 {
-                    requestTelemetry.Properties[prop.Key] = prop.Value;
+                    if (!requestTelemetry.Properties.ContainsKey(prop.Key))
+                    {
+                        requestTelemetry.Properties[prop.Key] = prop.Value;
+                    }
                 }
-            }
 
-            this.client.InitializeInstrumentationKey(requestTelemetry);
+                this.client.InitializeInstrumentationKey(requestTelemetry);
 
-            requestTelemetry.Source = GetAppIdFromRequestHeader(httpContext.Request.Headers, requestTelemetry.Context.InstrumentationKey);
+                requestTelemetry.Source = GetAppIdFromRequestHeader(httpContext.Request.Headers, requestTelemetry.Context.InstrumentationKey);
```

Comment:
do we need it on proactively sampled out telemetry?

Reply:
Sounds like a bug: with the new logic we only reply back with AppId if it's not null on incoming request. If I keep `GetAppIdFromRequestHeader()` only for sampled in telemetry, I have a chance or not setting AppId back on requests this particular service samples out (the other one may have decided differently due to its own sample rate).

It looks like I'd need to split check for request-context header (`IfExists()`) and then I can keep setting `Source` only for the telemetry I sample in. Does it sound right?

Reply:
Kept source for sampled out, upon investigation it does exactly one check for header existence and no op if it's not there. Moving this out will be equivalent perf.